oci_store.go

```diff
@@ -0,0 +1,138 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ratify
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"slices"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/oci"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/registry"
+)
+
+// OCIStore is a store implementation based on OCI image layout.
+type OCIStore struct {
+	name  string
+	store *oci.ReadOnlyStore
+}
+
+// NewOCIStoreFromFS creates a new [OCIStore] from the given filesystem in OCI
+// image layout.
+func NewOCIStoreFromFS(ctx context.Context, name string, fsys fs.FS) (*OCIStore, error) {
+	store, err := oci.NewFromFS(ctx, fsys)
+	if err != nil {
+		return nil, err
+	}
+	return &OCIStore{
+		name:  name,
+		store: store,
+	}, nil
+}
+
+// NewOCIStoreFromTar creates a new [OCIStore] from the given tarball in OCI
+// image layout.
+func NewOCIStoreFromTar(ctx context.Context, name string, path string) (*OCIStore, error) {
+	store, err := oci.NewFromTar(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	return &OCIStore{
+		name:  name,
+		store: store,
+	}, nil
+}
+
+// Name is the name of the store.
+func (s *OCIStore) Name() string {
+	return s.name
+}
+
+// Resolve resolves to a descriptor for the given artifact reference.
+func (s *OCIStore) Resolve(ctx context.Context, ref string) (ocispec.Descriptor, error) {
+	desc, err := s.store.Resolve(ctx, ref)
+	if err != nil {
+		if !errors.Is(err, errdef.ErrNotFound) {
+			return ocispec.Descriptor{}, err
+		}
+
+		// if there is no exact match on the full reference, try to resolve the
+		// tag / digest only.
+		reference, refErr := registry.ParseReference(ref)
+		if refErr == nil && reference.Reference != "" {
+			return s.store.Resolve(ctx, reference.Reference)
+		}
+		return ocispec.Descriptor{}, err
+	}
+	return desc, nil
+}
+
+// ListReferrers returns the immediate set of supply chain artifacts for the
+// given subject reference.
+func (s *OCIStore) ListReferrers(ctx context.Context, ref string, artifactTypes []string, fn func(referrers []ocispec.Descriptor) error) error {
+	desc, err := s.Resolve(ctx, ref)
+	if err != nil {
+		// return empty referrer list if the subject is not found.
+		// reference: https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#listing-referrers
+		if errors.Is(err, errdef.ErrNotFound) {
+			return nil
+		}
+		return err
+	}
+
+	referrers, err := registry.Referrers(ctx, s.store, desc, "")
```

Comment:
should we be concerned if there are too many referrers?

Reply:
It is fine since the file system is already in our trust boundary.